blackbird_python/blackbird/tests/test_listener.py

```diff
@@ -0,0 +1,464 @@
+
+        with monkeypatch.context() as m:
+            m.setattr("blackbird.blackbirdParser.blackbirdParser.ProgramContext.device", device)
+            walker.walk(bb, tree)
+
+        assert np.all(bb.var['A'] == np.array([[-1.0+1.0j, 2.7e5+0.2e-5j], [-0.1-2j, 0.2-0.1j]]))
+
+    def test_invalid_expression_type(self, parser, device, monkeypatch):
+        """Test exception is raised if the expression variable type is incorrect"""
+        text = "int Beta = -0.231e-6+5.21e-2j"
+        parser = parser(text)
+        tree = parser.start()
+
+        bb = BlackbirdListener()
+        walker = antlr4.ParseTreeWalker()
+
+        with monkeypatch.context() as m:
+            m.setattr("blackbird.blackbirdParser.blackbirdParser.ProgramContext.device", device)
+            with pytest.raises(TypeError, match=r"not of declared type int"):
+                walker.walk(bb, tree)
+
+    def test_invalid_array_type(self, parser, device, monkeypatch):
+        """Test exception is raised if the array variable type is incorrect"""
+        text = "float array A =\n\t-1.0+1.0j, 2.7e5+0.2e-5j\n\t-0.1-2j, 0.2-0.1j"
+        parser = parser(text)
+        tree = parser.start()
+
+        bb = BlackbirdListener()
+        walker = antlr4.ParseTreeWalker()
+
+        with monkeypatch.context() as m:
+            m.setattr("blackbird.blackbirdParser.blackbirdParser.ProgramContext.device", device)
+            with pytest.raises(TypeError, match=r"not of declared type float"):
+                walker.walk(bb, tree)
+
+    def test_invalid_array_shape(self, parser, device, monkeypatch):
+        """Test exception is raised if the array variable shape is incorrect"""
+        text = "complex array A[1, 2] =\n\t-1.0+1.0j, 2.7e5+0.2e-5j\n\t-0.1-2j, 0.2-0.1j"
+        parser = parser(text)
+        tree = parser.start()
+
+        bb = BlackbirdListener()
+        walker = antlr4.ParseTreeWalker()
+
+        with monkeypatch.context() as m:
+            m.setattr("blackbird.blackbirdParser.blackbirdParser.ProgramContext.device", device)
+            with pytest.raises(TypeError, match=r"has declared shape \(1, 2\) but actual shape \(2, 2\)"):
+                walker.walk(bb, tree)
+
+    def test_variable_expression(self, parser, device, monkeypatch):
+        """Test that a variable expression is correctly parsed"""
+        text = "float alpha = 0.32\nfloat gamma = (2.0*cos(alpha*pi)+1)**2"
+        parser = parser(text)
+        tree = parser.start()
+
+        bb = BlackbirdListener()
+        walker = antlr4.ParseTreeWalker()
+
+        with monkeypatch.context() as m:
+            m.setattr("blackbird.blackbirdParser.blackbirdParser.ProgramContext.device", device)
+            walker.walk(bb, tree)
+
+        assert bb.var['gamma'] == (2.0*np.cos(0.32*np.pi)+1)**2
+
+    def test_array_variable_expression(self, parser, device, monkeypatch):
+        """Test that a variable expression containing arrays is correctly parsed"""
```

Comment:
I thought it was stated in the docs that no array manipulations were allowed? Is scalar multiplication the only exception?

Reply:
Right, I meant no matrix-specific operations like matrix exp, matrix multiplication, power, etc. Element-wise operations are fine, and occur by default, so I thought may as well leave them in. I'll make this clearer in the docs.